File: abc017/d.cpp
#include<iostream>
#include<map>
using namespace std;

int N, M, aji[5000];

long long dfs(int n, int x)
{
  if(n >= N)
    return 1;
  if(n > N)
    return 0;
  int m = n;
  map<int, int> mem;
  long long res = 0;
  while(m < N && mem.find(aji[m]) == mem.end())
  {
    mem[aji[m]] = m;
    m++;
    res += dfs(m, x + 1);
  }

  return res % 1000000007;
}

int main()
{
  cin >> N >> M;
  for(int i = 0; i < N; i++)
    cin >> aji[i];
  cout << dfs(0, 0) << endl;
}

File: abc017/d.py
N, M = map(int, input().split())
aji = [0] + [int(input()) for i in range(N)]
aji[0] = aji[2]

ca = 1
dp = [1] + [0] * N
l = 0
mem = [0] * (M + 1)
for i, v in enumerate(dp):
    if i == 0:
        continue
    if l != mem[aji[i]]:
        l = mem[aji[i]]
        ca = 0
        for j in range(l, i):
            dp[i] = (dp[i] + dp[j]) % 1000000007
    else:
        dp[i] = ca
    ca += dp[i]
print(dp[-1])
